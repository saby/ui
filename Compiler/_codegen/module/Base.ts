/**
 */

import {
    IDependenciesController,
    createDependenciesController,
} from './Dependencies';
import { IExportsController, createExportsController } from './Exports';

/**
 * Интерфейс процессора, который выполняет компиляцию модуля в различный формат.
 */
export interface IModuleProcessor {
    /**
     * Установить имя компилируемого модуля.
     * @param name {string} Имя компилируемого модуля, которое содержит имя плагина.
     */
    setModuleName(name: string): void;

    /**
     * Установить флаг строго режима для компилируемого модуля.
     * @param value {boolean} Значение флага строго режима.
     */
    setStrictMode(value: boolean): void;

    /**
     * Добавить зависимость модуля. Если не указано имя зависимости, то зависимость считается неименованной.
     * @param ref {string} Путь зависимости.
     * @param name {string?} Имя зависимости.
     */
    addDependency(ref: string, name?: string): void;

    /**
     * Добавить блок код к телу компилируемого модуля.
     * @param block {string} Блок кода, который составляет тело модуля.
     */
    addCodeBlock(block: string): void;

    /**
     * Установить блок кода, который будет возвращен из модуля как экспортируемые данные.
     * @param block {string} Блок кода как экспортируемые данные.
     */
    setReturnableExport(block: string): void;

    /**
     * Выполнить компиляцию модуля.
     * @returns {string} JavaScript код модуля.
     */
    compile(): string;
}

/**
 * Абстрактный базовый класс процессора модулей.
 * Предоставляет общие данные и методы для выполнения компиляции.
 */
export abstract class Base implements IModuleProcessor {
    /**
     * Имя компилируемого модуля.
     * @protected
     */
    protected name: string;

    /**
     * Флаг включения строго режима для компилируемого модуля.
     * @protected
     */
    protected useStrict: boolean;

    /**
     * Массив кода, который составляет тело будущего модуля.
     * @protected
     */
    protected code: string[];

    /**
     * Инстанс контроллера зависимостей, необходимый для работы с именованными и
     * безымянными (анонимными) зависимостями модуля.
     * @protected
     */
    protected dependenciesController: IDependenciesController;

    /**
     * Инстанс контроллера экспортируемых данных.
     * @protected
     */
    protected exportsController: IExportsController;

    /**
     * Инициализировать инстанс базового класса.
     * @protected
     */
    protected constructor() {
        this.name = '';
        this.useStrict = false;
        this.code = [];
        this.dependenciesController = createDependenciesController();
        this.exportsController = createExportsController();
    }

    /**
     * Установить имя компилируемого модуля.
     * @param name {string} Имя компилируемого модуля, которое содержит имя плагина.
     */
    setModuleName(name: string): void {
        this.name = name;
    }

    /**
     * Установить флаг строго режима для компилируемого модуля.
     * @param value {boolean} Значение флага строго режима.
     */
    setStrictMode(value: boolean): void {
        this.useStrict = value;
    }

    /**
     * Добавить зависимость модуля. Если не указано имя зависимости, то зависимость считается неименованной.
     * @param ref {string} Путь зависимости.
     * @param name {string?} Имя зависимости.
     */
    addDependency(ref: string, name?: string): void {
        this.dependenciesController.addDependency(ref, name);
    }

    /**
     * Добавить блок код к телу компилируемого модуля.
     * @param block {string} Блок кода, который составляет тело модуля.
     */
    addCodeBlock(block: string): void {
        this.code.push(block);
    }

    /**
     * Установить блок кода, который будет возвращен из модуля как экспортируемые данные.
     * @param block {string} Блок кода как экспортируемые данные.
     */
    setReturnableExport(block: string): void {
        this.exportsController.setReturnableExport(block);
    }

    /**
     * Выполнить компиляцию модуля.
     * @abstract
     * @returns {string} JavaScript код модуля.
     */
    abstract compile(): string;
}
