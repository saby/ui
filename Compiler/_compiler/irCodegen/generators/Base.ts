/**
 * @author Krylov M.A.
 *
 * Базовый генератор обращений к именованным сущностям.
 */

import type { TPrimitive, IFormatter } from '../Interface';
import type { TAliasesMap } from '../Aliases';

import { wrapSequence } from '../types/String';

export default class Base {
    protected constructor(
        protected readonly formatter: IFormatter,
        protected readonly instance: string,
        protected readonly aliases: TAliasesMap,
        protected readonly isRelease: boolean
    ) { }

    protected getPropertyName(fn: string): string {
        if (this.isRelease) {
            if (this.aliases.has(fn)) {
                return `.${this.aliases.get(fn)}/*${fn}*/`;
            }
        }

        return `.${fn}`;
    }

    protected generateMethodCall(method: string, args: TPrimitive[]): string {
        return `${this.instance}${this.getPropertyName(method)}(${wrapSequence(args)})`;
    }

    protected generateFormattedMethodCall(method: string, args: TPrimitive[]): string {
        return `${this.instance}${this.getPropertyName(method)}(${this.formatter.formatSequence(args, 1)})`;
    }
}
