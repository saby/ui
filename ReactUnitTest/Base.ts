import { Control } from 'UICore/Base';
import { IControlOptions } from 'UICommon/Base';

export interface ITestOptions extends IControlOptions {
    testName?: string;
    afterMountCallback?: Function;
    afterUpdateCallback?: Function;
}

export class TestBaseControl<
    IOptions extends ITestOptions = ITestOptions
> extends Control<IOptions> {
    testName: string = '';
    afterMountCallback: Function = null;
    afterUpdateCallback: Function = null;
    _beforeMount(options: ITestOptions): void {
        this.afterMountCallback = options.afterMountCallback;
        this.afterUpdateCallback = options.afterUpdateCallback;
        this.testName = options.testName;
    }
    _afterMount(): void {
        this.afterMountCallback?.(this);
    }
    _afterUpdate(): void {
        this.afterUpdateCallback?.(this);
    }
}
