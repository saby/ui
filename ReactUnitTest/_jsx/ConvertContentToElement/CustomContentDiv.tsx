import InnerReactWithContent from './InnerReactWithContent';

export default function CustomContentDiv(): JSX.Element {
    return (
        <div className="customContentDiv">
            <InnerReactWithContent customContent={<div className="divContent" />} />
        </div>
    );
}
