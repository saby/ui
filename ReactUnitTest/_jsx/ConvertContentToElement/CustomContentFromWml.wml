<div class="customContentFromWml">
    <ReactUnitTest._jsx.ConvertContentToElement.InnerReactWithContent>
        <ws:customContent>
            <div class="divContent">divContent</div>
        </ws:customContent>
    </ReactUnitTest._jsx.ConvertContentToElement.InnerReactWithContent>

    <ReactUnitTest._jsx.ConvertContentToElement.InnerReactWithContent>
        <ws:customContent>
            <ReactUnitTest._jsx.ConvertContentToElement.SimpleWasaby attr:class="wasabyContent" />
        </ws:customContent>
    </ReactUnitTest._jsx.ConvertContentToElement.InnerReactWithContent>

    <ReactUnitTest._jsx.ConvertContentToElement.InnerReactWithContent>
        <ws:customContent>
            <ReactUnitTest._jsx.ConvertContentToElement.ClassNamed attr:class="reactContent" />
        </ws:customContent>
    </ReactUnitTest._jsx.ConvertContentToElement.InnerReactWithContent>
</div>
