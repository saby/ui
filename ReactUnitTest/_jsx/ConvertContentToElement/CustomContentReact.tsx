import InnerReactWithContent from './InnerReactWithContent';
import ClassNamed from './ClassNamed';

export default function CustomContentReact(): JSX.Element {
    return (
        <div className="customContentReact">
            <InnerReactWithContent customContent={ClassNamed} />
        </div>
    );
}
