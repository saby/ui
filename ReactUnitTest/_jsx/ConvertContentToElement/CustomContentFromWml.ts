import * as template from 'wml!ReactUnitTest/_jsx/ConvertContentToElement/CustomContentFromWml';
import { Control, TemplateFunction } from 'UI/Base';

export default class CustomContentFromWml extends Control {
    protected _template: TemplateFunction = template;
}
