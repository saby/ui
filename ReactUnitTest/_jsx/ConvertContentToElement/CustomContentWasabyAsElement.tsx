import * as WasabyWml from 'wml!ReactUnitTest/_jsx/ConvertContentToElement/SimpleWasaby';
import SimpleWasaby from './SimpleWasaby';
import InnerReactWithContent from './InnerReactWithContent';

export default function CustomContentWasabyAsElement(): JSX.Element {
    return (
        <div className="customContentWasabyAsElement">
            <InnerReactWithContent customContent={<WasabyWml className="wmlContent" />} />
            <InnerReactWithContent customContent={<SimpleWasaby className="controlContent" />} />
        </div>
    );
}
