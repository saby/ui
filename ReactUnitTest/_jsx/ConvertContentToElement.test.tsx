/**
 * @jest-environment jsdom
 */
import { unmountComponentAtNode, render } from 'react-dom';
import { act } from 'react-dom/test-utils';
import { Control } from 'UI/Base';
import CustomContentFromWml from './ConvertContentToElement/CustomContentFromWml';
import CustomContentDiv from './ConvertContentToElement/CustomContentDiv';
import CustomContentWasaby from './ConvertContentToElement/CustomContentWasaby';
import CustomContentWasabyAsElement from './ConvertContentToElement/CustomContentWasabyAsElement';
import CustomContentReact from './ConvertContentToElement/CustomContentReact';
import CustomContentReactAsElement from './ConvertContentToElement/CustomContentReactAsElement';

describe('ConvertContentToElement', () => {
    let container: HTMLDivElement;

    beforeEach(() => {
        container = document.createElement('div');
        document.body.appendChild(container);
    });

    afterEach(() => {
        unmountComponentAtNode(container);
        container.remove();
    });

    describe('Задание контента в wml', () => {
        it('Контентная опция', () => {
            act(() => {
                Control.createControl(CustomContentFromWml, {}, container);
            });
            expect(container).toMatchSnapshot();
        });
    });

    describe('Задание контента в tsx', () => {
        it('Wasaby как компонент', () => {
            act(() => {
                render(<CustomContentWasaby />, container);
            });
            expect(container).toMatchSnapshot();
        });
        it('Wasaby как реакт элемент', () => {
            act(() => {
                render(<CustomContentWasabyAsElement />, container);
            });
            expect(container).toMatchSnapshot();
        });
        it('DOM элемент', () => {
            act(() => {
                render(<CustomContentDiv />, container);
            });
            expect(container).toMatchSnapshot();
        });
        it('Чистый реакт компонент', () => {
            act(() => {
                render(<CustomContentReact />, container);
            });
            expect(container).toMatchSnapshot();
        });
        it('Чистый реакт компонент как реакт элемент', () => {
            act(() => {
                render(<CustomContentReactAsElement />, container);
            });
            expect(container).toMatchSnapshot();
        });
    });
});
