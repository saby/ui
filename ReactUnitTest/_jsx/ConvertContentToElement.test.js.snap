// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConvertContentToElement Задание контента в tsx DOM элемент 1`] = `
<div>
  <div
    class="customContentDiv"
  >
    <div
      class="InnerReactWithContent"
    >
      <div
        class="converterUserClassName divContent"
      />
    </div>
  </div>
</div>
`;

exports[`ConvertContentToElement Задание контента в tsx Wasaby как компонент 1`] = `
<div>
  <div
    class="customContentWasaby"
  >
    <div
      class="InnerReactWithContent"
    >
      <div
        class="simpleWasaby converterUserClassName"
        tabindex="0"
      >
        SimpleWasaby
      </div>
    </div>
    <div
      class="InnerReactWithContent"
    >
      <div
        class="simpleWasaby converterUserClassName"
        tabindex="0"
      >
        SimpleWasaby
      </div>
    </div>
  </div>
</div>
`;

exports[`ConvertContentToElement Задание контента в tsx Wasaby как реакт элемент 1`] = `
<div>
  <div
    class="customContentWasabyAsElement"
  >
    <div
      class="InnerReactWithContent"
    >
      <div
        class="simpleWasaby wmlContent converterUserClassName"
        tabindex="0"
      >
        SimpleWasaby
      </div>
    </div>
    <div
      class="InnerReactWithContent"
    >
      <div
        class="simpleWasaby controlContent converterUserClassName"
        tabindex="0"
      >
        SimpleWasaby
      </div>
    </div>
  </div>
</div>
`;

exports[`ConvertContentToElement Задание контента в tsx Чистый реакт компонент 1`] = `
<div>
  <div
    class="customContentReact"
  >
    <div
      class="InnerReactWithContent"
    >
      <div
        class="classNamed converterUserClassName "
      >
        ClassNamed
      </div>
    </div>
  </div>
</div>
`;

exports[`ConvertContentToElement Задание контента в tsx Чистый реакт компонент как реакт элемент 1`] = `
<div>
  <div
    class="customContentReactAsElement"
  >
    <div
      class="InnerReactWithContent"
    >
      <div
        class="classNamed converterUserClassName reactComponent"
      >
        ClassNamed
      </div>
    </div>
  </div>
</div>
`;

exports[`ConvertContentToElement Задание контента в wml Контентная опция 1`] = `
<div
  style="height: 100%; width: 100%; display: contents;"
>
  <div
    class="customContentFromWml"
  >
    <div
      class="InnerReactWithContent"
    >
      <div
        class="divContent converterUserClassName"
        tabindex="0"
      >
        divContent
      </div>
    </div>
    <div
      class="InnerReactWithContent"
    >
      <div
        class="simpleWasaby wasabyContent converterUserClassName"
        tabindex="0"
      >
        SimpleWasaby
      </div>
    </div>
    <div
      class="InnerReactWithContent"
    >
      <div
        class="classNamed reactContent converterUserClassName"
      >
        ClassNamed
      </div>
    </div>
  </div>
</div>
`;
